### ISSUE TITLE
Improvements to IRODS API

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#1639]
**Note: this is in error as stated below.** 

@hyi @pkdash @dtarb 
Much of the code for folders currently only works in iRODS, while the developers are mostly working in systems without iRODS. Thus using this code without change will **break the local development environment** even if it works in production. The fact that we need to run in local storage as well as iRODS means that we need a common storage API that can be implemented for both kinds of storage. That was previously django_irods, but that has been ignored in favor of calling irods directly because updating it is too awkward. Meanwhile, IrodsStorage, which started out as a class to mimic the interface in LocalStorage, now has plenty of extra functions that aren't needed for that mimicry. These functions need to be elsewhere. 

I suggest the following strategy: 

* Merge django_irods from a separate repository into HydroShare. Its existence as a separate repo isn't warranted. It is time to redevelop it.  
* Build HydroIrodsStorage as a subclass of IrodsStorage, having the functions we need for folders and other things specific to HydroShare, e.g., coupling between files and Django. 
* Build -- in parallel -- HydroLocalStorage as a subclass of DefaultStorage, having the functions we need for running in the local storage environment. 

That way, the local dev environment won't fail from use of iRODS-only functions. And it will be very clear that when we develop an iRODS function, we need to mirror that function in the local storage environment as well.

**Comment by hyi** [comment]
@alvacouch @pkdash @dtarb @mjstealey I am looping @mjstealey in for more details as needed but I can confirm that all developers work in their local development environments with iRODS as backend. It is set up in local_settings.py with default hsproxy iRODS user account and @mjstealey has scripts to clean up iRODS backend regularly set up in Jenkins. So iRODS backend storage is the only storage backend for HydroShare. We used to have a default local storage without iRODS a long long time ago, but when I implemented on-demand iRODS bag download and moved bagit operation from Django to iRODS side, that default local storage for development purposes could no longer be supported. Want to get this point across first before I answer your other questions.

**Comment by alvacouch** [comment]
What exactly am I binding to when I run a local instance? Does it make up a new iRODS instance?

**Comment by hyi** [comment]
@alvacouch If you don't run local irods docker containers by going to ```irods``` sub-directory and run the script to bring up irods docker containers in your local development environment, you are bound to a default hsproxy iRODS account in hydroshare test iRODS zone as set up in local_settings.py. So in your local development environment, you can either use the default hsproxy iRODS account set up or use local iRODS docker containers by running an extra script before running hsctl script. Looping @mjstealey in if you have more questions since @mjstealey set this all up.

**Comment by alvacouch** [comment]
I am currently having a lot of trouble getting that running, even on the develop branch. @mjstealey has been contacted already. 

Meanwhile, another basic question I have is why iRODS specific code is spread out in several files rather than being in a specific filesystem API?

**Comment by hyi** [comment]
@alvacouch @pkdash @dtarb @mjstealey Agreed IrodsStorage class has more methods that required as a sub-class of django ```Storage``` class, however, those methods are all storage-related and could be justified to stay there. Agreed it'd be more convenient to bring it into HydroShare repo rather than as a submodule, but will defer that action to @mjstealey. Most iRODS-related code I am aware of in other files/modules are getting an IrodsStorage object with different iRODS session objects associated to do iRODS storage related operations by calling APIs on IrodsStorage class. I may also called run method on the session object directly in a couple of places for folder-related operations just for simplicity since those operations are quite specific as I remembered and I thought a direct session run (a one line statement) is probably better than putting it into a method which is not likely used in other places for HydroShare. There might be room for improvement, but it has been working well so far and I have not found motivation to refactor it in some way.

**Comment by alvacouch** [comment]
What I want to do soon is to ensure that the iRODS code is in one place. I found the following: 

```
1132:    session.runCmd("iinit");
1133:    session.runCmd('iget', [ name, 'tempfile.' + name ])
```
... in `./hs_core/views/__init__.py`

```
551:    istorage.session.run("ibun", None, '-cDzip', '-f', output_zip_full_path, res_coll_input)
592:    istorage.session.run("ibun", None, '-xDzip', zip_with_full_path, unzip_path)
617:    istorage.session.run("imkdir", None, '-p', coll_path)
```
... in `./hs_core/views/utils.py`

```
786:        istorage.session.run("imkdir", None, '-p', res_contents_dir)
```
... in `./hs_core/hydroshare/utils.py`

These really need to be API calls. I don't suggest putting them into IrodsStorage, but rather, into a subclass HydroIrodsStorage, because there is a small chance of name collision in IrodsStorage if Django changes.

**Comment by alvacouch** [comment]
I also need to **extend** this functionality and need to be able to be sure that I am not duplicating something, and other people need to be able to find what I wrote.

**Comment by alvacouch** [comment]
Also consider the following code: 
```
hydro@hs-dev-1:~/hydroshare$ find . -name \*.py -exec grep -n DefaultStorage {} \; -print
22:from django.core.files.storage import DefaultStorage
1504:                                     else DefaultStorage())
1514:                                             settings, 'USE_IRODS', False) else DefaultStorage())
```
... in `./hs_core/models.py`

```
22:from django.core.files.storage import DefaultStorage
907:    return IrodsStorage() if getattr(settings, 'USE_IRODS', False) else DefaultStorage()
```
... in `./hs_core/hydroshare/utils.py`

**Comment by hyi** [comment]
@alvacouch The irods session runCmd calls in ```hs_core/views/__init__.py``` in ```get_file()``` method has been there since the beginning when I started to work on HydroShare and I don't think that method is ever used in our codebase. I suggest to remove that method since it is not used and we already have corresponding API method ```getFile()``` in IrodsStorage class. For other direct calls in views/utils.py, mainly, ```imkdir``` and ```ibun``` calls, I agree it is better to move them into API calls in IrodsStorage or a subclass of IrodsStorage. To avoid potential name collision, perhaps we can prefix the method with ```irods``` or ```hs``` as appropriate. Since IrodsStorage is already a subclass of Storage class, I don't see a need to make another hierarchy to create another subclass of IrodsStorage. I think that will complicate things a bit without the benefit of avoiding potential name collision, since Storage is still a parent class of HydroIrodsStorage and name collision is not allowed in HydroIrodsStorage either, right?

**Comment by hyi** [comment]
@alvacouch The leftover ```DefaultStorage```-related code can be removed now that we don't use it at all since USE_IRODS setting is always set to True.

**Comment by alvacouch** [comment]
That code is what led me to believe that DefaultStorage was still being used! So perhaps it would be good to get rid of it!

**Comment by hyi** [comment]
@alvacouch Agreed, now that you are at it, you can safely remove it I believe.

**Comment by hyi** [comment]
@alvacouch Agreed that if A is a subclass of B, then B's names are used in A but A's names are not utilized in B. However, in our case, you are proposing three classes, i.e., A (representing HydroIrodsStorage) is a subclass of B (representing IrodsStorage) which is a subclass of C (representing Storage), right? In this case, C's names will be used in A as well as in B, right? That is why I think making another level of subclass would not avoid name clash and probably prefixing our custom methods in IrodsStorage is the simplest way to make things work without potential risk.

**Comment by alvacouch** [comment]
There is another form of clash. In this kind, in the future Django decides to use a new name, say "foo". We've used that name already. Since methods in Storage refer to method in Storage (and not to methods in IrodsStorage) and since methods in HydroIrodsStorage refer to methods there and above, the only reason that the name clash would mean anything is if we wanted to use the function. Then we could use super() to get to it. The point is that it is really nice to keep the issues of 

1. Interfacing storage to Django and
2. Interfacing storage to HydroShare

somewhat separate, and the subclassing strategy doesn't really cost anything and makes that distinction clear.

**Comment by hyi** [comment]
I agree it is nice to keep interfacing storage to Django and interfacing storage to HydroShare separate via another layer of subclassing, but I think the risk for name clashing would be the same for both implementation strategies although this potential risk is small. I guess it is always safer to uniquely name our custom methods in subclasses to minimize potential clashing, prefixing is one simple way to do it.

**Comment by alvacouch** [comment]
@hyi If we are using IrodsStorage correctly, it mainly affects file pickers, e.g., FileField and ImageField. Meanwhile, we are using the special methods in IrodsStorage ourselves, but there is no way they would be relevant to FileField or ImageField.  It makes much more sense and is much more clean to specifically tailor IrodsStorage to FileField, and then put the methods we need for our own purposes elsewhere. 

So I must disagree. We need to concentrate on what the class was designed to do. It was designed to enable file picking. The routines we added to it get in the way of understanding this basic mission. They should be elsewhere. Since they must use the routines in IrodsStorage as their basis, it is sensible to put them in a subclass.

**Comment by hyi** [comment]
@alvacouch I can see the benefit of clean separation in terms of code readability and structure by putting additional helper routines in a subclass of IrodsStorage rather than into subclass IrodsStorage itself, but I view it as a nice-to-have rather than a must-to-have since I can see the argument on the other side as well for simplicity and direct use of instance variable without another layer of indirection introduced by subclassing. So if you want to improve the code structure by introducing another layer of subclassing, I am not in objection at all, and by all means, please go for it as long as you have time to do this refactoring.

**Comment by alvacouch** [comment]
@pkdash @hyi @dtarb Unlike many of you, I have an extra responsibility, which is to assure that the code base is ready for production use in Jan at CUASHI. This means that I need to assure that the code is understandable and maintainable by CUAHSI developers. Any complexity that remains **becomes my responsibility to explain**. I consider this refactoring essential for that reason, including: 

1. Protecting the functions in `IrodsStorage` and assuring the cohesion of that module for a particular purpose of enabling `FileField` functions. 
2. Avoiding violations of encapsulation by putting iRODS-related functions into a suitable class, and avoiding embedding generic `.run(*)` functions into other places in the code separate from the API. 

These might not be essential from the point of view of basic function, but in my opinion -- as senior architect of CUAHSI -- they **are** essential to the ability of anyone else to make sense of the code and be able to maintain it and troubleshoot problems. 

Thus, my revised proposal is as follows: refactor iRODS code so that: 

1. Only FileField related functions appear in IrodsStorage. 
2. A new class "HydroStorage" subclasses IrodsStorage and contains
   a. Existing functions in IrodsStorage that are not part of the Django Storage API (thus ensuring the    
      cohesiveness of IrodsStorage as a Django module) 
   b. Functions needed to remove the direct use of `.run(*)` outside the HydroStorage module, thus insuring its encapsulation.

**Comment by pkdash** [comment]
@alvacouch Your proposal for this change seems good to me. After AGU I would also like to take some time do some refactoring/restructuring rather than adding new features.

**Comment by pkdash** [comment]
@alvacouch Is this still an issue?
